pdd/PDD-multistream.md

```diff
@@ -0,0 +1,139 @@
+
+The multistream protocol does not cover:
+
+- discovering participants, selecting transports, and establishing connections
+- managing the state of the connection
+
+multistream enables several types of streams to be used over one single stream, like an intelligent message broker that offers the ability to negotiate the protocol and version that is going to be used. To simplify, a visual representation can be:
+
+```
+┌ ─ ─ ─ ─ ─ ─ ┌ ─ ─ ─ ─ ─ ─ ┐┌ ─ ┐
+ dht-id/1.0.1│ bitswap/1.2.3  ...
+└ ─ ─ ─ ─ ─ ─ └ ─ ─ ─ ─ ─ ─ ┘└ ─ ┘
+┌────────────────────────────────┐
+│ multistream-select             │
+└────────────────────────────────┘
+┌────────────────────────────────┐
+│transport                       │
+│  [TCP, UDP, uTP, ...]          │
+└────────────────────────────────┘
+```
+
+multistream doesn't cover stream multiplexing over the same connection, however, we can achieve this by leveraging functionality offered by SPDY or HTTP/2.
+
+```
+┌─── ───┐┌─── ───┐┌ ┌─── ───┐┌─── ───┐┌┐
+ a/1.0.0  b/1.0.0  │ a/1.0.0  b/1.0.0 ││
+└─── ───┘└─── ───┘└ └─── ───┘└─── ───┘ ┘
+┌─── ──── ──── ──── ┌─── ──── ──── ──── ┌───
+│multistream-select││multistream-select││...│
+└ ──── ──── ──── ──┘└ ──── ──── ──── ──┘└ ──┘
+┌─── ──── ──── ──── ──── ──── ──── ──── ──── ┌───
+│stream multiplexing                        ││...│
+│ [HTTP/2, SPDY]                            ││   │
+└─ ──── ──── ──── ──── ──── ──── ──── ──── ─┘└─ ─┘
+┌────────────────────────────────────────────────┐
+│ multistream-select                             │
+└────────────────────────────────────────────────┘
+┌────────────────────────────────────────────────┐
+│transport                                       │
+│  [TCP, UDP, uTP, ...]                          │
+└────────────────────────────────────────────────┘
+```
+
+multistream messages have the following structure:
+
+```
+┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─
+ ┌───────────────┐┌───────┐│
+││varint with    ││message│
+ │ message length││       ││
+│└───────────────┘└───────┘
+ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘
+```
+
+However, for readability reasons, we will omit the `varint` part on the Compliance Spec.
+
+
+Other reference material and discussion:
+- https://github.com/ipfs/node-ipfs/issues/13#issuecomment-109802818
+
+## Protocol Compliance Tests Spec
+
+Given the protocol spec, an implementation of the multistream-select protocol has to comply with the following scenarios:
+
+#### 1 - push-stream
+
+In a push-stream example (one-way stream), we have two agents:
+
+- 'broadcast' - where messages are emited 
+- 'silent' - listener to the messages emited by the broadcast counterparty
+
+Compliance test 1 (human readable format, without varint):
+```
+# With a connection established between silent - broadcast
+< /multistream/1.0.0       # multistream header
+  < /bird/3.2.1              # Protocol header
+    < hey, how is it going?    # First protocol message
+```
```

Comment:
On Fri, Jun 12, 2015 at 04:45:57PM -0700, Juan Batiz-Benet wrote:
> > +# With a connection established between silent - broadcast
> > +< /multistream/1.0.0       # multistream header
> > +  < /bird/3.2.1              # Protocol header
> > +    < hey, how is it going?    # First protocol message

This is the multistream-select view (because of the nested streams),
but for conversations like this where the streams are strictly nested
(e.g., no going back to an ancestor stream after saying something in a
child stream), you could also run this test against a non-multiplexed
(uniplexed?) multistream implementation.  Elaborating on the
test-harness changes for the multiplexed transport:

1. Spin up connection to implementation, attaching the
   multistream-select driver.  We probably want a
   ‘/multistream-select/{version}’ handshake (or optimistic
   broadcast?) to make sure we're speaking the same multistream-select
   language (or at least can figure out what went wrong in a
   brodcast-to-silent situation).
2. Send the ‘/multistream/1.0.0’ message through the
   multistream-select driver.
3. Get a new multiplexed stream (I'm not sure which agent is
   responsible for creating the new stream) and attach the
   multistream-1.0.0 driver.
4. Send the ‘/bird/3.2.1’ message through the multistream driver.
5. Get a new multiplexed stream and attach the bird-3.2.1 driver.
6. Send the ‘hey, how is it going?’ message through the bird driver.
7. Teardown streams?

While for the uniplexed transport, we'd have:

1. Spin up connection to implementation, attaching the
   multistream-select driver.  We probably want a
   ‘/multistream-select/{version}’ handshake (or optimistic
   broadcast?) to make sure we're speaking the same multistream-select
   language (or at least can figure out what went wrong in a
   brodcast-to-silent situation).
2. Send the ‘/multistream/1.0.0’ message through the
   multistream-select driver.
3. Disconnect the multistream-select driver and attach the
   multistream-1.0.0 driver to the stream.
4. Send the ‘/bird/3.2.1’ message through the multistream driver.
5. Disconnect the multistream driver and attach the bird-3.2.1 driver
   to the stream.
6. Send the ‘hey, how is it going?’ message through the bird driver.
7. Teardown stream?

The test-suite's transport driver should have a IsMultiplexed() check
so it can decide which of these approaches to take (or so it can error
out if you ask it to run a multiplex-requiring transcript over a
uniplexed transport).